### PR TITLE
Secure bindings

### PR DESCRIPTION
**Pull request by the-djmaze**
This change prevents access outside the $context scope.
It means that `window.*` properties, functions and classes are not accessible to `data-bind=""`.

This does it by eliminating one `with` statement and wrap the context into a Proxy.

You could loosen this code to your liking by changing it to something like:
```
get: (target, key) => target[key] || target['$data'][key] || window[key]
```
Or with custom get callbacks or other restrictions.

For example it would block: 
```html
<span data-bind="text: document.cookie"`></span>
```

As a side-effect this should also improve render speed, but i couldn't test it properly.

### CODE REVIEW

**Comment by justlep** [comment]
Hi, thanks for your PR and the work you put in it. I'm sorry I can't merge it though, for a couple of reasons:

1. It breaks 250 browser tests (`spec/runner.html`).
   See: https://github.com/justlep/knockout-esnext/#changes-testingdebugging

2. It produces inconsistent results:

    ```html
    <body>
        <h2 data-bind="text: typeof document"></h2>
        <h2 data-bind="text: typeof document"></h2>
        <h2 data-bind="text: typeof document"></h2>
    </body>
    ```

    outputs:
    ```
    undefined
    object
    object
    ```

3. Your proxy getter 
    ```javascript
    get: (target, key) => target[key] || target['$data'][key]
    ```
    is not equivalent to 
    ```javascript
    with(target) {
      with(target.$data) {
        ...
    ```
    in case of falsy values.

4. It's a breaking change that vastly limits the flexibility of binding strings,
breaking existing projects, cutting compatibility with original KnockoutJS.

5. It adds little improvement on security because
   - the only malicious/misbehaving `data-bind` attribute values I can think of are either added by intention or through http injection (i.e. "game over" either way)
   - cookies that not intened to be read by frontend scripts should be HttpOnly in the first place (though I am aware cookies were probably just an example for global data here)
   
6. Performance hit from creating loads of proxies, even more from accessing vm properties through proxy-getters.
This quick test here shows 2-3x longer timings with proxy: 
https://gist.github.com/justlep/12a9609891cb8366a9934d7b1e54de40

7. `bindingContext['$element'] = node` looks prone to memory leaks, but would need investigation

Bottom line is, I see why some reactive libraries went the Proxy road, but I'm not convinced Knockout would benefit from them.

**Comment by the-djmaze** [comment]
I understand your concerns and thanks for testing them.

`with` is [deprecated](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with) and from there i used the proxy idea.

There is still 1 deprecated `with`, but to eliminate that, it would mean that all templates need to be changed (or the binding parser).
As most `data-bind="href: xIn"` need to become `data-bind="href: $data.xIn"`


3. you are correct. this changes the behavior. I also see the code should be reverse anyway `target['$data'][key] || target[key]`

4. is definitely true. That is why i opted `target['$data'][key] || target[key] || window[key]`
This could be made an option.

5. Yes, cookie was just an example.

6. I've changed your gist to be more in line with Knockout and Proxy is still 2x slower. It is just depends how secure something has to be.
https://jsfiddle.net/ost3Lp6n/

7. could be. i've just removed $element in my code as i never use it.
